pennylane/gradients/gradient_transform.py

```diff
@@ -114,70 +114,94 @@ def assert_no_tape_batching(tape, transform_name):
+    is not provided, all trainable parameters are considered.
+
+    Args:
+        tape (`~.QuantumScript`): the tape to analyze
+        argnum (int, list(int), None): Indices for trainable parameters(s)
+            to compute the Jacobian for.
+
+    Returns:
+        list: list of the trainable parameter indices
+
+    """
+
+    if argnum is None:
+        return [idx for idx, _ in enumerate(tape.trainable_params)]
+
+    if isinstance(argnum, int):
+        argnum = [argnum]
+
+    if len(argnum) == 0:
+        warnings.warn(
+            "No trainable parameters were specified for computing the Jacobian.",
+            UserWarning,
+        )
+
+    return argnum
+
+
+def _try_zero_grad_from_graph_or_get_grad_method(tape, param_index, use_graph=True):
+    """Gets the gradient method of a parameter. If use_graph=True, analyze the
+    circuit graph to find if the parameter has zero gradient.
+
+    Args:
+        tape (`~.QuantumScript`): the tape to analyze
+        param_index (int): the index of the parameter to analyze
+        use_graph (bool): whether to use the circuit graph to find if
+            a parameter has zero gradient
+
     """
+
     # pylint:disable=protected-access
-    if grad_fn is not None:
-        if getattr(tape, "_gradient_fn", None) is grad_fn:
-            # gradient analysis has already been performed on this tape
-            return
+    par_info = tape._par_info[param_index]
 
-        tape._gradient_fn = grad_fn
+    if use_graph:
+        op_or_mp = tape[par_info["op_idx"]]
+        if not any(tape.graph.has_path(op_or_mp, mp) for mp in tape.measurements):
+            # there is no influence of this operation on any of the observables
+            return "0"
 
-    for idx, info in enumerate(tape._par_info):
-        if idx not in tape.trainable_params:
-            # non-trainable parameters do not require a grad_method
-            info["grad_method"] = None
-        else:
-            par_info = tape._par_info[idx]
-            op = par_info["op"]
+    return par_info["op"].grad_method
 
-            if not qml.operation.has_grad_method(op):
-                # no differentiation method is registered for this operation
-                info["grad_method"] = None
 
-            elif (tape._graph is not None) or use_graph:
-                # if `op` is an observable, we want the associated MP that's in the graph
-                op_or_mp = tape[par_info["op_idx"]]
-                if not any(tape.graph.has_path(op_or_mp, mp) for mp in tape.measurements):
-                    # there is no influence of this operation on any of the observables
-                    info["grad_method"] = "0"
-                    continue
+def _find_gradient_methods(tape, trainable_param_indices, use_graph=True):
+    """Returns a dictionary with gradient information of each trainable parameter."""
 
-            info["grad_method"] = op.grad_method
+    return {
+        idx: _try_zero_grad_from_graph_or_get_grad_method(

```

Comment:
If we can get rid of `use_graph`, we can have a one line code instead of two private functions